matlab_simulation/05-estimation/ekf/ekf_example.m

```diff
@@ -0,0 +1,98 @@
+% Extended Kalman filter example
+%
+% Radar measurement of airplane position while flying at a constant
+% altitude and velocity. In this case the motion model is linear, but the
+% measurement model is nonlinear.
+%
+% This example can be found on slide 111 of Estimation-I in the notes.
+clear;clc;
+
+%% Create AVI object
+makemovie = 1;
+if(makemovie)
+    vidObj = VideoWriter('ekf.avi');
+    vidObj.Quality = 100;
+    vidObj.FrameRate = 8;
+    open(vidObj);
+end
+
+% Discrete time step
+dt = 0.1;
+
+% Initial State
+x0 = [20 -2 3]';
+
+% Prior
+mu = [22 -1.8 6.5]'; % mean (mu)
+S = 1*eye(3);% covariance (Sigma)
+
+% Discrete motion model
+Ad = [ 1 dt 0 ; 0 1 0; 0 0 1];
+
+R = [.0001 0 0; 0 .0001 0 ; 0 0 .0001];
+[RE, Re] = eig (R);
+
+% Measurement model defined below
+Q = [.0001];
+
+% Simulation Initializations
+Tf = 10;
+T = 0:dt:Tf;
+n = length(Ad(1,:));
+x = zeros(n,length(T));
+x(:,1) = x0;
+m = length(Q(:,1));
+y = zeros(m,length(T));
+
+% Variables to store during iteration
+mup_S = zeros(n,length(T));
+mu_S = zeros(n,length(T));
+K_S = zeros(n,n,length(T));
+
+%% Main loop
+for t=2:length(T)
+    %%% Simulation
+    % Select a motion disturbance
+    e = RE*sqrt(Re)*randn(n,1);
+    % Update state
+    x(:,t) = Ad*x(:,t-1) + e;
+
+    % Take measurement
+    % Select a motion disturbance
+    d = sqrt(Q)*randn(m,1);
+    % Determine measurement
+    y(:,t) = sqrt(x(1,t)^2 + x(3,t)^2) + d;
+
+
+    %%% Extended Kalman Filter Estimation
+    [mu, S, mup, K] = ekf(mu, S, y(:,t), ...
+                          @airplane_radar_motion_model, ...
+                          @airplane_radar_measurement_model, ...
+                          @airplane_radar_linearized_motion_model, ...
+                          @airplane_radar_linearized_measurement_model, ...
+                          Q, R);
+
+    % Store results
+    mup_S(:,t) = mup;
+    mu_S(:,t) = mu;
+    K_S(:,t) = K;
+
+
+    %%% Plot results
```

Comment:
Good call. Unfortunately Matlab is super slow if you have a legend during the video, since it redraws it every time. I've now made it so the legends/labels show up in just the last frame so it goes at a reasonable speed.